rustler_codegen/src/context.rs

```diff
@@ -44,11 +45,37 @@ impl<'a> Context<'a> {
             _ => panic!("Struct can only have one lifetime argument"),
         };
 
+        let type_param_idents = ast
+            .generics
+            .type_params()
+            .map(|type_param| {
+                if !type_param.attrs.is_empty() {
+                    panic!("Attributes for type parameters are currently not supported.");
+                }
+                if !type_param.bounds.is_empty() {
+                    panic!("Type parameter bounds are currently not supported.");
+                }
+                if type_param.eq_token.is_some() {
+                    panic!("Type parameter constraints are currently not supported.");
+                }
+                if type_param.default.is_some() {
+                    panic!("Type parameter defaults are currently not supported.");
+                }
+                &type_param.ident
+            })
+            .collect::<Vec<_>>();
+
         let ident = &ast.ident;
-        let ident_with_lifetime = if has_lifetime {
-            quote! { #ident <'a> }
+
+        let lifetimes = if has_lifetime {
+            vec![quote! {'a}]
```

Comment:
Is the restriction on one lifetime still necessary? It would be nice if we simply allowed an arbitrary number of lifetimes.

Reply:
Maybe we could then write something like this:

```rust
let lifetimes: Vec<_> = ast.generics.lifetimes().collect();
let ident_with_generics = quote! { #ident < #(#lifetimes,)* #(#type_param_idents),* > };
```

Reply:
Lifetimes are harder, because the decoder trait has a lifetime as well. Note that this lifetime is set equal to the one of the datatype, if it has one. I haven't understood how it interacts with the generated code completely. When there are multiple lifetimes, we might need to alpha-rename one of them to prevent a name clash.

But I'll happily give it a go in a subsequent PR if that's fine for you!